src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -42,15 +47,68 @@ template <typename T>
 using Partition = std::conditional_t<std::is_trivially_destructible_v<T>, uninitialized_vector<PartitionedElement<T>>,
                                      std::vector<PartitionedElement<T>>>;
 
-// The small_vector holds the first n values in local storage and only resorts to heap storage after that. 1 is chosen
-// as n because in many cases, we join on primary key attributes where by definition we have only one match on the
-// smaller side.
-using SmallPosList = boost::container::small_vector<RowID, 1>;
+// Stores the mapping from HashedType to positions. Conceptually, this is similar to an (unordered_)multimap, but it
+// has some optimizations for the performance-critical probe() method. Instead of storing the matches directly in the
+// hashmap (think map<HashedType, PosList>), we store an offset. This keeps the hashmap small and makes it easier to
+// cache.
+template <typename HashedType>
+class PosHashTable {
+  // In case we consider runtime to be more relevant, the flat hash map performs better (measured to be mostly on par
+  // with bytell hash map and in some cases up to 5% faster) but is significantly larger than the bytell hash map.
+  //
+  // The hash table stores the relative offset of a SmallPosList (see below) in the vector of SmallPosLists. The range
+  // of the offset does not limit the number of rows in the partition but the number of distinct values. If we end up
+  // with a partition that has more values, the partitioning algorithm is at fault.
+  using Offset = uint32_t;
+  using HashTable = ska::bytell_hash_map<HashedType, Offset>;
```

Comment:
Not sure I can follow. Do you mean if the indirection costs us anything if the hash map has a size of 1?

Reply:
Before find() yielded the PosList directly, right? Now, you are using the offset to get the PosList from somewhere else.

Can that lead to a regression? I’d guess it can.

Reply:
I would argue that the benefit from the smaller hashmap greatly outweighs that indirection. Do you have an idea on how to measure that edge case?

Reply:
I'll post a benchmark before lunch.

Reply:
Benchmarks this code on nemea (GCC 8.3, numactl-pinned on one node): https://gist.github.com/Bouncner/e38d2b4984d86c751489c908a44a5b54

Current master:
```
------------------------------------------------------------------------------------------------
Benchmark                                                         Time           CPU Iterations
------------------------------------------------------------------------------------------------
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          66609664 ns   66609809 ns         11
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          66710450 ns   66711962 ns         11
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          66263806 ns   66255242 ns         11
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          67445408 ns   67431171 ns         11
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          67076748 ns   67054165 ns         11
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST_mean     66821215 ns   66812470 ns         11
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST_median   66710450 ns   66711962 ns         11
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST_stddev     453498 ns     448019 ns         11
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          27412393 ns    2234384 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          26238070 ns    2183268 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          27522159 ns    2206834 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          29690635 ns    2072070 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          32688117 ns    2077772 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT_mean     28710275 ns    2154866 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT_median   27522159 ns    2183268 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT_stddev    2549130 ns      75215 ns        100
```

This PR:
```
------------------------------------------------------------------------------------------------
Benchmark                                                         Time           CPU Iterations
------------------------------------------------------------------------------------------------
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          57964365 ns   57965739 ns         12
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          57991902 ns   57990718 ns         12
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          57773908 ns   57769177 ns         12
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          59011539 ns   59004609 ns         12
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST          58907330 ns   58895396 ns         12
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST_mean     58329809 ns   58325128 ns         12
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST_median   57991902 ns   57990718 ns         12
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_ST_stddev     582032 ns     578138 ns         12
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          26697230 ns    1774228 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          26718147 ns    1755765 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          25758581 ns    1772984 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          26285396 ns    1785634 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT          26344461 ns    1770514 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT_mean     26360763 ns    1771825 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT_median   26344461 ns    1772984 ns        100
TPCHDataMicroBenchmarkFixture/BM_Mrks_HashJoin_MT_stddev     390353 ns      10694 ns        100
```

Reply:
So we can resolve this?